src/directives/scaleselector.js

```diff
@@ -111,14 +111,18 @@ ngeo.ScaleselectorController = function($scope, $element, $attrs) {
       ($scope.$eval(scalesExpr));
   goog.asserts.assert(this.scales !== undefined);
 
-  var zoomLevels = Object.keys(this.scales).map(Number);
-  zoomLevels.sort(ol.array.numberSafeCompareFunction);
-
   /**
    * @type {Array.<number>}
    * @export
    */
-  this.zoomLevels = zoomLevels;
+  this.zoomLevels;
+
+  $scope.$watchCollection(goog.bind(function() {
```

Comment:
As long as I don't have the same array length with different keys, your propsal works fine. 
As I don't have such usecase, it looks good to me.
I push the new version.